### PR TITLE
Don't allow test revisions that conflict with built in cfgs

### LINKED ISSUE #128964: compiletest: we should not allow revision names that collide with builtin cfg names
Example: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/error.3A.20unexpected.20.60--cfg.20windows.60.20flag/near/459869068

```
--- stderr -------------------------------
error: unexpected `--cfg windows` flag
  |
  = note: config `windows` is only supposed to be controlled by `--target`
  = note: manually setting a built-in cfg can and does create incoherent behaviors
  = note: `#[deny(explicit_builtin_cfgs_in_flags)]` on by default

error: aborting due to 1 previous error
```

compiletest will try to set a `--cfg <revision_name>` that matches the revision name, but this goes poorly if the `<revision_name>` collides with a builtin cfg. Builtin cfgs can be found at https://github.com/rust-lang/rust/blob/c9bd03cb724e13cca96ad320733046cbdb16fbbe/compiler/rustc_session/src/config/cfg.rs#L90, but we should also deny `test` revision name because it has special meaning `cfg(test)`.

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #128964

Sorry @heysujal I started working on this about 1 minute before your comment by complete coincidence 😅

### CODE REVIEW

File: src/tools/compiletest/src/common.rs [hunk 1]
```diff
@@ -376,6 +376,7 @@ pub struct Config {
     pub only_modified: bool,
 
     pub target_cfgs: OnceLock<TargetCfgs>,
+    pub builtin_cfgs: OnceLock<Vec<String>>,
```
Comment (jieyouxu) [inline src/tools/compiletest/src/common.rs:379]: Remark: I feel like this is kinda overkill, but I'll see if this is even needed during the rework

File: src/tools/compiletest/src/common.rs [hunk 2]
```diff
@@ -651,6 +657,17 @@ pub enum Endian {
     Big,
 }
 
+fn builtin_cfgs(config: &Config) -> Vec<String> {
```
Comment (jieyouxu) [inline src/tools/compiletest/src/common.rs:660]: Suggestion: `HashSet<String>` or maybe run a `dedup`? There's a bunch of duplicates when I run that on a nightly rustc, doesn't matter that much for the purpose of this check 🤷 

```rs
panic="abort"
panic="unwind"
proc_macro
relocation_model="dynamic-no-pic"
relocation_model="pic"
relocation_model="pie"
relocation_model="ropi"
relocation_model="ropi-rwpi"
relocation_model="rwpi"
relocation_model="static"
rustfmt
sanitize="address"
sanitize="cfi"
```

**Review by jieyouxu — COMMENTED** [review]
Thanks, this LGTM with one suggestion.

**Comment by heysujal** [comment]
Hi @clubby789, its fine. The issue didn't had any comment of anyone working on it, so I thought its up for grabs. 
Can you explain your approach of fixing this? 
I looked at your changes and want to understand how did you figure out where and what changes to make.

**Review by jieyouxu — APPROVED** [review]
Thanks, you can r=me once PR CI is green.

**Comment by clubby789** [comment]
>  Can you explain your approach of fixing this?
@heysujal 
As suggested by Urgau on the original issue, we can use `--print=check-cfg` to gather all builtin cfg names. When running a test with a revision, we check if that revision is in the list and panic with an error if it is. Otherwise, we continue with the old behaviour of adding a `--cfg` argument for it

**Comment by Urgau** [comment]
This is now/already covered by the deny-by-default [`explicit_builtin_cfgs_in_flags`](https://doc.rust-lang.org/nightly/rustc/lints/listing/deny-by-default.html#explicit-builtin-cfgs-in-flags) lint.
No need to add an extra check in compiletest for that.

*Also the `test` cfg is [a bit special](https://github.com/rust-lang/rust/blob/c926476d013fbb2ca43bd5259d0a7228009a9cb2/compiler/rustc_session/src/config/cfg.rs#L111-L113).*

**Comment by clubby789** [comment]
I could see that error getting easily lost in a compile-fail test. test could make sense to remove though

**Comment by jieyouxu** [comment]
> This is already covered by the deny-by-default [`explicit_builtin_cfgs_in_flags`](https://doc.rust-lang.org/nightly/rustc/lints/listing/deny-by-default.html#explicit-builtin-cfgs-in-flags) lint.
> 
> What is the advantage over the native check in rustc?

The preset handling in compiletest is such that revisions are handled as early props but compile flagd are handled at late props. Revisions set cfgs of revision names implicitly and its very confusing for the end user seeing the lint fire only when the test is about to run, I would like revision names that share same name as builtin cfg names to get rejected earlier so it's obvious it's an invalid revision name.

File: src/tools/compiletest/src/common.rs [hunk 3]
```diff
@@ -651,6 +657,17 @@ pub enum Endian {
...
 }
 
+fn builtin_cfg_names(config: &Config) -> HashSet<String> {
+    rustc_output(
+        config,
+        &["--print=check-cfg", "-Zunstable-options", "--check-cfg=cfg()"],
+        Default::default(),
+    )
+    .lines()
+    .map(|l| if let Some((name, _)) = l.split_once('=') { name.to_string() } else { l.to_string() })
```
Comment (Urgau) [inline src/tools/compiletest/src/common.rs:667]: Just a FWIW, this parsing is quite primitive, it doesn't handle the [`--print=check-cfg` specifics](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/print-check-cfg.html#printcheck-cfg), fortunately this is only collecting names and shouldn't be in any specific case so it should always work.

Comment (jieyouxu) [reply src/tools/compiletest/src/common.rs:667]: Yeah, this is only querying for the cfg name, otherwise this can't be used

**Comment by jieyouxu** [comment]
Also @Urgau good point about test, let's ban that for compiletest even if its not part of builtin cfgs anymore.

**Pull request merged**